### ISSUE TITLE
testsuite reports failure for GEMM for Randomize vectors and matrices using powers of 2 in narrow precision range

### ISSUE THREAD

**Issue by kvaragan** [flame/blis#413]
We get error failures as:
Output:
blis_sgemm_cn_ccc                    1     1     1     0.00       -nan   FAILURE
blis_cgemm1m_hc_ccc                  1     1     1     0.00       -nan   FAILURE
blis_dgemm_nc_ccc                    1     1     1     0.00       -nan   FAILURE
blis_dgemm_cc_ccc                    1     1     1     0.00       -nan   FAILURE
blis_dgemm_ct_ccc                    1     1     1     0.00       -nan   FAILURE
blis_dgemm_ht_ccc                    1     1     1     0.00       -nan   FAILURE
blis_dgemm_hh_ccc                    1     1     1     0.00       -nan   FAILURE
blis_zgemm1m_tc_ccc                  1     1     1     0.00       -nan   FAILURE

In input.general we enable 1 to use Randomize vectors and matrices using powers of 2 in narrow precision range.
However this error doesn't occur when use real values [-1,1].

**Comment by devinamatthews** [comment]
@kvaragan can you explain a little more how the initialization is done (or @fgvanzee if this is something built-in)?

**Comment by kvaragan** [comment]
Initialization is built into testsuite.

**Comment by devinamatthews** [comment]
@fgvanzee ?

**Comment by fgvanzee** [comment]
@kvaragan Thanks for this bug report, Kiran. I've isolated the issue. 

The problem is that the current implementation of both scalar randomization functions (regular or narrow powers of 2) can sometimes return 0. For the narrow powers of 2, this happens *much* more frequently because only a handful of values are possible. Why is this of concern? When I wrote the narrow powers of 2 randomizer, I figured that 0 should be a valid value in our test matrices just like any other value in the range. The problem is that it can sometimes cause the numerical test to malfunction because it will attempt to normalize by the largest value in the matrix, even if it's 0. (So the nans were presumably there result of computing 0/0.) This can very easily happen for 1x1 matrices when using the narrow powers of 2 randomization.

So, in retrospect, this problem *can* manifest for randomization on the real range [-1,1], but never does in practice because it is highly improbable.

At first I was going to simply disallow 0 as a valid random value--likely for both types of randomization--but then I realized that the problem is not zeroes, per se, but rather objects that are *completely* zero. So instead, I think I will check the randomized object's 1-norm, and if it's zero, re-randomize until it's not zero.
